pkg/login/social/social.go

```diff
@@ -225,3 +233,97 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
 
 	return result
 }
+
+func GetOAuthHttpClient(name string) (*http.Client, error) {
+	name = strings.TrimPrefix(name, "oauth_")
+
+	info, ok := setting.OAuthService.OAuthInfos[name]
+	if !ok {
+		return nil, fmt.Errorf("Could not find %s in OAuth Settings", name)
+	}
+
+	// handle call back
+	tr := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: info.TlsSkipVerify,
+		},
+	}
+	oauthClient := &http.Client{
+		Transport: tr,
+	}
+
+	if info.TlsClientCert != "" || info.TlsClientKey != "" {
+		cert, err := tls.LoadX509KeyPair(info.TlsClientCert, info.TlsClientKey)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to setup TlsClientCert error: %s", err)
+		}
+
+		tr.TLSClientConfig.Certificates = append(tr.TLSClientConfig.Certificates, cert)
+	}
+
+	if info.TlsClientCa != "" {
+		caCert, err := ioutil.ReadFile(info.TlsClientCa)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to setup TlsClientCa error: %s", err)
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+		tr.TLSClientConfig.RootCAs = caCertPool
+	}
+	return oauthClient, nil
+}
+
+var GetCurrentOAuthToken = func(ctx context.Context, user models.SignedInUser) (*oauth2.Token, error) {
+	authInfoQuery := &models.GetAuthInfoQuery{UserId: user.UserId}
+	if err := bus.Dispatch(authInfoQuery); err != nil {
```

Comment:
Would the same code be fine if it is in the new oauthtoken package as you suggested?

If GetCurrentOAuthToken (in oauthtoken, not social) has to take this as an argument, then we will have the database code twice -- once in datasource_plugin_wrapper_v2.go and once in ds_proxy.go.

Reply:
Yes, parts of it. Would make sense to have database interactions in oauthtoken package. oauthtoken package should use the social package for retrieving the token (and refresh if needed) since the social package has all the internal knowledge of providers, http clients etc. Basically the oauthtoken package will be the coordinator for getting a token, but reach out to other packages for getting the user token from database, refresh it, if needed, and store it in database if refreshed. 

Does this make sense?